private/protocol/eventstream/debug.go

```diff
@@ -0,0 +1,132 @@
+package eventstream
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"strconv"
+)
+
+type decodedMessage struct {
+	rawMessage
+	Headers decodedHeaders `json:"headers"`
+}
+type jsonMessage struct {
+	Length     json.Number    `json:"total_length"`
+	HeadersLen json.Number    `json:"headers_length"`
+	PreludeCRC json.Number    `json:"prelude_crc"`
+	Headers    decodedHeaders `json:"headers"`
+	Payload    []byte         `json:"payload"`
+	CRC        json.Number    `json:"message_crc"`
+}
+
+func (d *decodedMessage) UnmarshalJSON(b []byte) error {
+	var jsonMsg jsonMessage
+	if err := json.Unmarshal(b, &jsonMsg); err != nil {
+		return err
+	}
+
+	d.Length = numAsUint32(jsonMsg.Length)
+	d.HeadersLen = numAsUint32(jsonMsg.HeadersLen)
+	d.PreludeCRC = numAsUint32(jsonMsg.PreludeCRC)
+	d.Headers = jsonMsg.Headers
+	d.Payload = jsonMsg.Payload
+	d.CRC = numAsUint32(jsonMsg.CRC)
+
+	return nil
+}
+
+func (d *decodedMessage) MarshalJSON() ([]byte, error) {
+	jsonMsg := jsonMessage{
+		Length:     json.Number(strconv.Itoa(int(d.Length))),
+		HeadersLen: json.Number(strconv.Itoa(int(d.HeadersLen))),
+		PreludeCRC: json.Number(strconv.Itoa(int(d.PreludeCRC))),
+		Headers:    d.Headers,
+		Payload:    d.Payload,
+		CRC:        json.Number(strconv.Itoa(int(d.CRC))),
+	}
+
+	return json.Marshal(jsonMsg)
+}
+
+func numAsUint32(n json.Number) uint32 {
+	v, err := n.Int64()
+	if err != nil {
+		panic(fmt.Sprintf("failed to get int64 json number, %v", err))
```

Comment:
probably don't want to return a serialization error as this is embedded within the eventstream's debug logger. This code is only used by the eventstream unit tests for decoding the eventstream test data files.  I can update it to return an error instead of panic though.